test/functional/p2p_blocksonly.py

```diff
@@ -69,18 +49,55 @@ def run_test(self):
         peer_2_info = self.nodes[0].getpeerinfo()[1]
         assert_equal(peer_2_info['whitelisted'], True)
         assert_equal(peer_2_info['permissions'], ['noban', 'forcerelay', 'relay', 'mempool', 'download'])
-        assert_equal(self.nodes[0].testmempoolaccept([sigtx])[0]['allowed'], True)
-        txid = self.nodes[0].testmempoolaccept([sigtx])[0]['txid']
+        assert_equal(self.nodes[0].testmempoolaccept([tx_hex])[0]['allowed'], True)
 
         self.log.info('Check that the tx from forcerelay first_peer is relayed to others (ie.second_peer)')
         with self.nodes[0].assert_debug_log(["received getdata"]):
-            first_peer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+            first_peer.send_message(msg_tx(tx))
             self.log.info('Check that the forcerelay peer is still connected after sending the transaction')
             assert_equal(first_peer.is_connected, True)
             second_peer.wait_for_tx(txid)
             assert_equal(self.nodes[0].getmempoolinfo()['size'], 1)
         self.log.info("Forcerelay peer's transaction is accepted and relayed")
 
+        self.nodes[0].disconnect_p2ps()
+        self.nodes[0].generate(1)
+
+    def blocks_relay_conn_tests(self):
+        self.log.info('Tests with node in normal mode with block-relay-only connections')
+        self.restart_node(0, ["-noblocksonly"])  # disables blocks only mode
+        assert_equal(self.nodes[0].getnetworkinfo()['localrelay'], True)
+
+        # Ensure we disconnect if a block-relay-only connection sends us a transaction
+        self.nodes[0].add_outbound_p2p_connection(P2PInterface(), connection_type="blockrelay")
+        assert_equal(self.nodes[0].getpeerinfo()[0]['relaytxes'], False)
+        _, txid, tx_hex = self.check_p2p_tx_violation(index=2)
+
+        self.log.info("Check that txs from RPC are not sent to blockrelay connection")
+        conn = self.nodes[0].add_outbound_p2p_connection(P2PTxInvStore(), connection_type="blockrelay")
+
+        assert_equal(self.nodes[0].testmempoolaccept([tx_hex])[0]['allowed'], True)
+        self.nodes[0].sendrawtransaction(tx_hex)
+        conn.sync_with_ping()
+        assert(txid not in conn.get_invs())
```

Comment:
This doesn't actually test that the transaction won't get announced by the node, since transaction announcements are batched and only sent when the `nNextInvSend` timer pops. The `sync_with_ping()` above doesn't guarantee that there's nothing in the `setInventoryTxToSend` waiting to be sent (ie I think this test would still pass if this were an outbound-full-relay peer). I don't have any good suggestions except maybe adding a comment. Testing for the absence of something is hard!

Reply:
ah, yeah I missed that. luckily `nNextInvSend` is mockable so I'm able to bump time forward to ensure the transaction would send. I also fixed a bug with the matching condition (was comparing different types). I think the test now functions correctly- if you remove the `connection_type="blockrelay"` param when adding the p2p conn, the tx sends and the test fails. what do you think?

Reply:
resolving this conversation so we can continue over here: https://github.com/bitcoin/bitcoin/pull/19315#discussion_r525413509